File: 2016/Day2.ts
import * as fs from "fs"


let keypad: string[][] =
[["1", "2", "3"],
 ["4", "5", "6"],
 ["7", "8", "9"]];

let cursor: number[] = [1, 1];

function moveCursor(line: string, cursor: number[]): number[] {
    for (const char of line.trim()) {
        let lastCursor: number[] = cursor.slice();
        switch (char){
            case "R":
                cursor[0]++;
                break;
            case "L":
                cursor[0]--;
                break;
            case "U":
                cursor[1]--;
                break;
            case "D":
                cursor[1]++;
                break;
        }
        if (cursor[0] < 0 || cursor[1] < 0 || cursor[0] > 2 || cursor[1] > 2) {
            cursor = lastCursor;
        }
    }
    return cursor;
}

function readCursor(cursor: number[]): string{
    return keypad[cursor[1]][cursor[0]];
}

// const data: string= `ULL
// RRDDD
// LURDL
// UUUUD`;


fs.readFile("inputs/Day2.txt", "utf8", (err, data) => {
    if (err) throw err;
let code: string = "";
for (const line of data.split("\n")) {
    cursor = moveCursor(line, cursor)
    code += readCursor(cursor);
}
console.log(code)
});

File: 2016py/Day2.py
keypad = [[1, 2, 3],
          [4, 5,6],
          [7,8,9]]
x = y = 1

def move_cursor(x, y, movement):
    if movement == "U" and y>0:
        y -= 1
    elif movement == "D" and y<2:
        y += 1
    elif movement == "L" and x>0:
        x -= 1
    elif movement == "R" and x<2:
        x += 1
    return x,y

def get_lines(in_str:str):
    return map(str.strip, in_str.strip().splitlines())


def main(x,y, lines):
    code = ""
    for line in lines:
        for i in line:
            x,y = move_cursor(x, y, i)
        code += str(keypad[y][x])
    return code





test_str = """ULL
RRDDD
LURDL
UUUUD
"""
with open("inputs/Day2.txt","r") as f:
    in_str = f.read()

if __name__ == "__main__":
    print(main(x, y,get_lines(in_str)))